## Extraccion de los datos desde la API

In [1]:
import pandas as pd
import requests

# Realizar la solicitud GET a la API
url = 'https://restcountries.com/v3.1/all'
response = requests.get(url)
data = response.json()

# Transformar el JSON en un DataFrame de Pandas
df = pd.json_normalize(data)

# Seleccionar las 8 columnas más importantes
important_columns = ['name.common', 'population', 'area',  'region',
                     'subregion', 'capital', 'borders', 'timezones']

df = df.loc[:, important_columns]

# Imprimir las primeras filas del DataFrame con las 10 columnas seleccionadas
print(df.head())

   name.common  population       area  region         subregion     capital  \
0        Spain    47351567   505992.0  Europe   Southern Europe    [Madrid]   
1  Afghanistan    40218234   652230.0    Asia     Southern Asia     [Kabul]   
2        Sudan    43849269  1886068.0  Africa   Northern Africa  [Khartoum]   
3     Bulgaria     6927288   110879.0  Europe  Southeast Europe     [Sofia]   
4      Germany    83240525   357114.0  Europe    Western Europe    [Berlin]   

                                         borders         timezones  
0                      [AND, FRA, GIB, PRT, MAR]  [UTC, UTC+01:00]  
1                 [IRN, PAK, TKM, UZB, TJK, CHN]       [UTC+04:30]  
2            [CAF, TCD, EGY, ERI, ETH, LBY, SSD]       [UTC+03:00]  
3                      [GRC, MKD, ROU, SRB, TUR]       [UTC+02:00]  
4  [AUT, BEL, CZE, DNK, FRA, LUX, NLD, POL, CHE]       [UTC+01:00]  


## LIMPIEZA DE LOS DATOS 

In [2]:
df

name.common  population       area  region         subregion  \
0          Spain    47351567   505992.0  Europe   Southern Europe   
1    Afghanistan    40218234   652230.0    Asia     Southern Asia   
2          Sudan    43849269  1886068.0  Africa   Northern Africa   
3       Bulgaria     6927288   110879.0  Europe  Southeast Europe   
4        Germany    83240525   357114.0  Europe    Western Europe   
..           ...         ...        ...     ...               ...   
245      Burundi    11890781    27834.0  Africa    Eastern Africa   
246      Senegal    16743930   196722.0  Africa    Western Africa   
247     Ethiopia   114963583  1104300.0  Africa    Eastern Africa   
248   Luxembourg      632275     2586.0  Europe    Western Europe   
249        China  1402112000  9706961.0    Asia      Eastern Asia   

           capital                                            borders  \
0         [Madrid]                          [AND, FRA, GIB, PRT, MAR]   
1          [Kabul]                     [IRN, PAK, TKM, UZB, TJK, CHN]   
2       [Khartoum]                [CAF, TCD, EGY, ERI, ETH, LBY, SSD]   
3          [Sofia]                          [GRC, MKD, ROU, SRB, TUR]   
4         [Berlin]      [AUT, BEL, CZE, DNK, FRA, LUX, NLD, POL, CHE]   
..             ...                                                ...   
245       [Gitega]                                    [COD, RWA, TZA]   
246        [Dakar]                          [GMB, GIN, GNB, MLI, MRT]   
247  [Addis Ababa]                     [DJI, ERI, KEN, SOM, SSD, SDN]   
248   [Luxembourg]                                    [BEL, FRA, DEU]   
249      [Beijing]  [AFG, BTN, MMR, HKG, IND, KAZ, NPL, PRK, KGZ, ...   

            timezones  
0    [UTC, UTC+01:00]  
1         [UTC+04:30]  
2         [UTC+03:00]  
3         [UTC+02:00]  
4         [UTC+01:00]  
..                ...  
245       [UTC+02:00]  
246             [UTC]  
247       [UTC+03:00]  
248       [UTC+01:00]  
249       [UTC+08:00]  

[250 rows x 8 columns]

In [3]:
df=df.rename(columns={'name.common':'Name'})

In [4]:
df['capital'] = df['capital'].astype(str)
df['borders'] = df['borders'].astype(str)
df['timezones'] = df['timezones'].astype(str)

In [5]:
df

Name  population       area  region         subregion  \
0          Spain    47351567   505992.0  Europe   Southern Europe   
1    Afghanistan    40218234   652230.0    Asia     Southern Asia   
2          Sudan    43849269  1886068.0  Africa   Northern Africa   
3       Bulgaria     6927288   110879.0  Europe  Southeast Europe   
4        Germany    83240525   357114.0  Europe    Western Europe   
..           ...         ...        ...     ...               ...   
245      Burundi    11890781    27834.0  Africa    Eastern Africa   
246      Senegal    16743930   196722.0  Africa    Western Africa   
247     Ethiopia   114963583  1104300.0  Africa    Eastern Africa   
248   Luxembourg      632275     2586.0  Europe    Western Europe   
249        China  1402112000  9706961.0    Asia      Eastern Asia   

             capital                                            borders  \
0         ['Madrid']                ['AND', 'FRA', 'GIB', 'PRT', 'MAR']   
1          ['Kabul']         ['IRN', 'PAK', 'TKM', 'UZB', 'TJK', 'CHN']   
2       ['Khartoum']  ['CAF', 'TCD', 'EGY', 'ERI', 'ETH', 'LBY', 'SSD']   
3          ['Sofia']                ['GRC', 'MKD', 'ROU', 'SRB', 'TUR']   
4         ['Berlin']  ['AUT', 'BEL', 'CZE', 'DNK', 'FRA', 'LUX', 'NL...   
..               ...                                                ...   
245       ['Gitega']                              ['COD', 'RWA', 'TZA']   
246        ['Dakar']                ['GMB', 'GIN', 'GNB', 'MLI', 'MRT']   
247  ['Addis Ababa']         ['DJI', 'ERI', 'KEN', 'SOM', 'SSD', 'SDN']   
248   ['Luxembourg']                              ['BEL', 'FRA', 'DEU']   
249      ['Beijing']  ['AFG', 'BTN', 'MMR', 'HKG', 'IND', 'KAZ', 'NP...   

                timezones  
0    ['UTC', 'UTC+01:00']  
1           ['UTC+04:30']  
2           ['UTC+03:00']  
3           ['UTC+02:00']  
4           ['UTC+01:00']  
..                    ...  
245         ['UTC+02:00']  
246               ['UTC']  
247         ['UTC+03:00']  
248         ['UTC+01:00']  
249         ['UTC+08:00']  

[250 rows x 8 columns]

## CREACION DE LA TABLA EN REDSHIFT

Create table STG_countries(

Name varchar(200),

population int,

area decimal(10,2),

region varchar(250),

subregion varchar(250),

capital varchar(250),

borders varchar(250),

timezones varchar(250)

)

## Login a REDSHIFT

In [16]:
!pip install sqlalchemy
!pip install psycopg2

In [6]:
import psycopg2


In [21]:
url="data-engineer-cluster.cyhh5bfevlmn.us-east-1.redshift.amazonaws.com"
data_base="data-engineer-database"
user="gastonrios32_coderhouse"
with open("C:/Users/grios/Documents/User_REDSHIFT.txt",'r') as f:
    pwd= f.read()


In [22]:
try:
    conn = psycopg2.connect(
        host='data-engineer-cluster.cyhh5bfevlmn.us-east-1.redshift.amazonaws.com',
        dbname=data_base,
        user=user,
        password=pwd,
        port='5439'
    )
    print("Connected to Redshift successfully!")
    
except Exception as e:
    print("Unable to connect to Redshift.")
    print(e)

Connected to Redshift successfully!


In [23]:
conn

<connection object at 0x0000026F5160AD00; dsn: 'user=gastonrios32_coderhouse password=xxx dbname=data-engineer-database host=data-engineer-cluster.cyhh5bfevlmn.us-east-1.redshift.amazonaws.com port=5439', closed: 0>

In [24]:
cur = conn.cursor()
# Execute a SQL query to select data from a table
cur.execute("SELECT * FROM STG_countries")
# Fetch the results
results = cur.fetchall()
results

[]

In [25]:
from psycopg2.extras import execute_values
cur = conn.cursor()
# Define the table name
table_name = 'gastonrios32_coderhouse.stg_countries'
# Define the columns you want to insert data into
columns = ['Name', 'population', 'area','region','subregion','capital','borders','timezones']
# Generate 
values = [tuple(x) for x in df.to_numpy()]
insert_sql = f"INSERT INTO {table_name} ({', '.join(columns)}) VALUES %s"
# Execute the INSERT statement using execute_values
cur.execute("BEGIN")
execute_values(cur, insert_sql, values)
cur.execute("COMMIT")
# Close the cursor and connection
#cur.close()
#conn.close()

In [26]:
cur.close()
conn.close()